---
title: "Human Genetics and Genomics Computational Biology Course"
author: "Dennis Hazelett & Simon Coetzee"
date: "Spring 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Course Abstract

The goal of this course is to take students on an in-depth tour of a single common bioinformatics analysis pipeline. We will deconstruct the RNA-seq analysis pipeline starting with the easiest, highest level visualizations and working our way back through successively more advanced techniques. We will start at the endpoint of the analysis, with an existing, finalized dataset, going step-by-step through the pipeline until we reach the very beginning with alignment and quality control steps. We will emphasize practical utility of the methods, plus exposure to important ideas and concepts for students to explore on their own at a later date. 

## Course Format

The course takes place over 6 weeks with 1.5 hour lectures on Tuesdays and practical workshops on Thursdays for a total of 12 lectures and workshops. Students should gain an understanding of what is involved with this type of analysis and be able to use their notes to recapitulate it with a future dataset if needed.

### Special considerations for COVID-19
Due to social distancing protocols the course will take place online via a password-locked Zoom session. Details to be distributed via email. 

## Expectations: Attendance, Homework & Grading

The course is worth 100 points. Homework is expected to be turned in on time. Late assignments will be given a maximum of 50% of full credit up to one lecture _after_ the original due date. Recommended readings are indicated on the syllabus. During the course, students will choose one or more transcription factors from the CRISPRi dataset and follow the lecture series in completing the steps of an analysis pipeline from data acquisition through through hypothesis testing and visualization. Each stage of analysis and results is completed in successive homework assignments ("Labs") or question sets. Each question set is worth 10 points and will be graded based on completeness. Each student will also be required to present their homework at least once during the course during a workshop. For this informal presentation, the student will be graded on their ability to discuss the figures they generated and any challenges encountered (10 points). A final exam worth 30 points will be given at the 12 lecture.

**Attendance is required** no exceptions. You may obtain permission ahead of time or with extenuating circumstances after the fact from the graduate school (email to Emma Yates Kassler). Homework assignments are required on time (see schedule below) regardless of attendance or for half credit **one lecture late**. Unexcused absences result in an incomplete grade.

### Lecture & assignment schedule:

#### Tuesday, May 5 _Introduction_
**Lecture topics:** Course structure and expectations. Install GENAVI. Availability of course materials. **Homework Assignment I:** (Lab) install genavi locally.

#### Thursday, May 7 _DGE_
**Lecture topics:** ENCODE datasource. TCGA data and TCGAbiolinksGUI. Preparing data for use in GENAVI. Differential gene expression (DGE) analysis using GENAVI. **Homework Asssignment II:** (Lab) Perform DGE on a TF of your choice.

#### Tuesday, May 12 _Review_
Workshop / Homework Review

#### Thursday, May 14 _Exploratory Analysis_
**Lecture topics:** Unsupervised clustering. Principle Components Analysis (PCA). Correlation. **Assignment III:** (Lab) Exploratory analysis of your chosen TFs in the CRISPRi dataset.

#### Tuesday, May 19 _Review_
Workshop / Homework Review

#### Thursday, May 21 _GO Analysis_
**Lecture topics:** Gene Ontology (GO) & Pathway Analysis, David, GoRilla, GENAVI. Gene Set Enrichment Analysis (GSEA) and MSigDB. Installing software. **Assignment IV:** (Lab) GO analysis of your TF DGE set.

#### Tuesday, May 26 _Review_
Workshop / Homework Review

#### Thursday, May 28 _NGS & File Formats_ (Coetzee)
**Lecture topics:** Log files from alignment and QC analyses. Structure and interpretation of NGS file formats ( _e.g._ FASTQ, .sam _etc._) **Assignment V:** (Question set)

#### Tuesday, Jun 2 _Review_
Workshop / Homework Review

#### Thursday, Jun 4 _MultiQC_ (Coetzee)
**Lecture topics:** Performing Quality control with FASTQC and MultiQC. MegaQC for aggregating and databasing MultiQC reports. **Assignment VI:** (Question Set)

#### Tuesday, Jun 9 _Overview_
**Lecture topics:** Question Set review. Course Overview / Review. Course Feedback and Evaluation.

#### Thursday, Jun 11 _Final Exam_

### Oral presentations

During one of the 3 scheduled workshop sessions students are expected **_on a volunteer basis_** to discuss the results of a homework assignment and any challenges encountered (worth 10 points).

### Schedule and Due Dates:

```{r schedule, echo=FALSE, warning=FALSE}
sched <- data.frame(
  day = c("Tue", "Thu", "Tue", "Thu", "Tue", "Thu", "Tue", "Thu", "Tue", "Thu", "Tue", "Thu"),
  date = format(as.Date(c("2020-05-05", "2020-05-07", "2020-05-12", "2020-05-14", "2020-05-19", "2020-05-21", "2020-05-26", "2020-05-28", "2020-06-02", "2020-06-04", "2020-06-09", "2020-06-11")), "%m/%d"),
  lecture = c("Intro", "DGE", "Wkshp", "Expl Anls", "Review", "GO Anls", "Wkshp", "Formats", "Wkshp", "MultiQC", "Review", "Exam"),
  hmwk = c("LI", "LII", "", "LIII", "", "LIV", "", "QSV", "", "QSVI", "", ""),
  due = c("", "LI", "LII", "", "LIII", "", "LIV", "", "QSV", "", "QSVI", "EI")
)
knitr::kable(sched, caption = "**Course schedule and outline.** Roman numerals are preceded by 'L' for labs, 'QS' for question sets and 'E' for exams. Optional times for oral presentation are during workshops.")
```

## Course Materials

All course materials, including this syllabus, lab workflows, question sets, and lecture slides are available online through the graduate school website or github.